### PR TITLE
Avoid saving empty strings as translations in case search

### PR DESCRIPTION
**Pull request by orangejenny**
## Product Description
https://dimagi-dev.atlassian.net/browse/USH-2335

## Technical Summary
[This change](https://github.com/dimagi/commcare-hq/pull/31871/files#r915186610) had the side effect of causing search properties without a hint to store an empty string in the hint attribute, e.g., `{default_lang: ""}` where it previously would have been an empty dictionary (and similarly for other localizable text properties).

For hints, this meant that [this check](https://github.com/dimagi/commcare-hq/blob/7f4796d4b51fcd98eb6050cf900081173912e49e/corehq/apps/app_manager/suite_xml/post_process/remote_requests.py#L265) started returning true even if there was no real hint text present. That then caused hints to be included in the suite.

Blank translations get replaced by [non-breaking spaces](https://github.com/dimagi/commcare-translations/blob/60e5ce47400d645eee332d8f1ea7a0d8db57767a/commcare_translations.py#L88-L90). Formplayer would pass the non-breaking space along to web apps, which would see a hint and display the hint popup, with no content except that non-breaking space.

This PR updates the save logic so that empty strings will not be stored in the translations dicts for case search properties.

## Feature Flag
Case search & claim

### Safety story
I'm basically relying on test coverage. Open to feedback.

### Automated test coverage

In PR.

### QA Plan

Not requesting QA.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/views/modules.py [hunk 1]
```diff
@@ -1009,16 +1009,27 @@ def _update_search_properties(module, search_properties, lang='en'):
     True
 
     """
+
+    # Replace translation for current language in a translations dict
+    def _update_translation(old_obj, new_data, attr):
```
Comment (esoergel) [inline corehq/apps/app_manager/views/modules.py]: Nice, I was also thinking a helper might be nice for this.

File: corehq/apps/app_manager/views/modules.py [hunk 2]
```diff
@@ -1009,36 +1009,47 @@ def _update_search_properties(module, search_properties, lang='en'):
...
     """
+
+    # Replace translation for current language in a translations dict
+    def _update_translation(old_obj, new_data, old_attr, new_attr=None):
+        new_attr = new_attr or old_attr
+        values = getattr(old_obj, old_attr) if old_obj else {}
+        try:
+            values.pop(lang)
+        except KeyError:
+            pass
```
Comment (esoergel) [inline corehq/apps/app_manager/views/modules.py:1020]: nit:
```suggestion
        values.pop(lang, None)
```

File: corehq/apps/app_manager/views/modules.py [hunk 3]
```diff
@@ -1009,36 +1009,47 @@ def _update_search_properties(module, search_properties, lang='en'):
     True
 
     """
+
+    # Replace translation for current language in a translations dict
+    def _update_translation(old_obj, new_data, old_attr, new_attr=None):
+        new_attr = new_attr or old_attr
```
Comment (esoergel) [inline corehq/apps/app_manager/views/modules.py:1015]: this is kind of confusing.  If I've got it right, it's:
* `old_attr`: the attr in `current`
* `new_attr`: the attr in `prop`?

I'll think about whether I can come up with anything more obvious, but nothing comes to mind offhand.

**Review by esoergel — APPROVED** [review]

File: corehq/apps/app_manager/views/modules.py [hunk 4]
```diff
@@ -1009,36 +1009,47 @@ def _update_search_properties(module, search_properties, lang='en'):
     True
 
     """
+
+    # Replace translation for current language in a translations dict
+    def _update_translation(old_obj, new_data, old_attr, new_attr=None):
```
Comment (esoergel) [inline corehq/apps/app_manager/views/modules.py:1014]: At it's core this is a way to join together the old and new dicts, so you could limit the scope of the function to just that - make it accept the old dict and the new translation, then handle the `pop`ing and joining here, but leave figuring out how to find the old dict and the new translation to the callers.  I am mildly annoyed by the need to guard against `current = None` everywhere, though.  Oh well, maybe there's not a perfectly clean option here.

File: corehq/apps/app_manager/views/modules.py [hunk 5]
```diff
@@ -1020,21 +1020,21 @@ def _update_search_properties(module, search_properties, lang='en'):
             'label': {**_current_label, lang: prop['label']},
             'hint': {**_current_hint, lang: prop['hint']},
         }
-        if prop['default_value']:
+        if prop.get('default_value'):
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/views/modules.py:1035]: Guessing this is so it's easier to test - seems like a fair trade-off

File: corehq/apps/app_manager/tests/test_modules.py [hunk 6]
```diff
@@ -32,6 +33,19 @@ def setUp(self):
         self.module.case_type = 'another_case_type'
         self.form = self.module.new_form("Untitled Form", None)
 
+    def test_update_search_properties(self):
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/tests/test_modules.py:37]: nothing, so far as I can tell 😲

File: corehq/apps/app_manager/tests/test_modules.py [hunk 7]
```diff
@@ -32,6 +34,57 @@ def setUp(self):
...
+        module.search_config.properties = [
+            CaseSearchProperty(name='name', label={'fr': 'Nom'}),
+            CaseSearchProperty(name='age', label={'fr': 'Âge'}),
+        ]
+        props = list(_update_search_properties(module, [
+            {'name': 'name', 'label': 'Name'},
+            {'name': 'dob', 'label': 'Date of birth'}
+        ], "en"))
+        self.assertEqual(props[0]['label'], {'en': 'Name', 'fr': 'Nom'})
+        self.assertEqual(props[1]['label'], {'en': 'Date of birth'})
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/tests/test_modules.py:48]: yeah, I stole this test from the [function comment](https://github.com/dimagi/commcare-hq/blob/629716fc7b54c2a153371589c11ed0d49f046ee2/corehq/apps/app_manager/views/modules.py#L991-L1015) - perhaps I should copy the comments, it wasn't apparent to me at first either

File: corehq/apps/app_manager/views/modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/views/modules.py:1032]: yeah, it spares me from needing to provide every attribute in test data

File: corehq/apps/app_manager/views/modules.py (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/app_manager/views/modules.py:1015]: yep, that's it, open to suggestions on the naming (and structure - I don't think this code is especially elegant, it's just good enough)

File: corehq/apps/app_manager/tests/test_modules.py [hunk 8]
```diff
@@ -32,6 +34,57 @@ def setUp(self):
...
+
+    def test_update_search_properties_blank(self):
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/tests/test_modules.py]: sure, 33b21c28362

**Pull request merged**